splunk_otel/profiling/__init__.py

```diff
@@ -44,161 +44,285 @@
+            profiling_logger.emit(log_record)
 
+            processing_time = time.time() - time_begin
+            wait_for = max(
+                call_stack_interval_seconds - processing_time, min_call_stack_interval_seconds
+            )
 
-def _encode_cpu_profile(thread_states, stacktraces, timestamp_unix_nanos, interval):
-    str_table = StringTable()
-    locations_table = OrderedDict()
-    functions_table = OrderedDict()
+            with self.condition:
+                self.condition.wait(wait_for)
 
-    def get_function(file_name, function_name):
-        key = f"{file_name}:{function_name}"
-        fun = functions_table.get(key)
+        # Hack around batch log processor getting stuck on application exits when the profiling endpoint is not reachable
+        # Could be related: https://github.com/open-telemetry/opentelemetry-python/issues/2284
+        # pylint: disable-next=protected-access
+        self.exporter._shutdown = True
+        self.logger_provider.shutdown()
+        with self.condition:
+            self.condition.notify_all()
+
+    def get_ignored_thread_ids(self):
+        if self.include_internal_stacks:
+            return []
+
+        ignored_ids = [threading.get_ident()]
+
+        log_processor_thread_id = self.extract_log_processor_thread_id()
+
+        if log_processor_thread_id:
+            ignored_ids.append(log_processor_thread_id)
+
+        return ignored_ids
+
+    def extract_log_processor_thread_id(self):
+        if hasattr(self.batch_processor, "_worker_thread"):
+            # pylint: disable-next=protected-access
+            worker_thread = self.batch_processor._worker_thread
+
+            if isinstance(worker_thread, threading.Thread):
+                return worker_thread.ident
+
+        return None
+
+    def to_log_record(self, stacktraces, timestamp_unix_nanos):
+        encoded = self.get_cpu_profile(stacktraces, timestamp_unix_nanos)
+        encoded_profile = base64.b64encode(encoded).decode()
+
+        return LogRecord(
+            timestamp=timestamp_unix_nanos,
+            trace_id=0,
+            span_id=0,
+            trace_flags=TraceFlags(0x01),
+            severity_number=SeverityNumber.UNSPECIFIED,
+            body=encoded_profile,
+            resource=self.resource,
+            attributes={
+                "profiling.data.format": "pprof-gzip-base64",
+                "profiling.data.type": "cpu",
+                "com.splunk.sourcetype": "otel.profiling",
+                "profiling.data.total.frame.count": len(stacktraces),
+            },
+        )
 
-        if fun is None:
-            name_id = str_table.index(function_name)
-            fun = profile_pb2.Function()
-            fun.id = len(functions_table) + 1
-            fun.name = name_id
-            fun.system_name = name_id
-            fun.filename = str_table.index(file_name)
-            functions_table[key] = fun
+    def get_cpu_profile(self, stacktraces, timestamp_unix_nanos):
```

Comment:
There's a lot going on in this function. As I mentioned above, we may want to refactor it and the function that creates the stacktraces into a separate class (or classes). I didn't want to increase the scope of this proposal so I have left it out of this PR.